lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,156 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+    auto moduleFileType =
+      file_types::lookupTypeForExtension(path::extension(depPath));
+    auto isSwiftmodule =
+      moduleFileType == file_types::TY_SwiftModuleFile;
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (!(isSwiftmodule || isSwiftinterface))
+      continue;
+
+    auto dep = pathToModuleDecl.find(depPath);
+    if (dep != pathToModuleDecl.end()) {
+      // Great, we recognize the path! Check if the file is still around.
+
+      ModuleDecl *depMod = dep->second;
+      if(depMod->isResilient() && !isSwiftinterface) {
+        SmallString<256> moduleAdjacentInterfacePath(depPath);
+        computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+        if (fs::real_path(moduleAdjacentInterfacePath, buffer, true)) {
+          err << "The module " << depMod->getName().str() << " has library"
+              << " evolution enabled but we're recording a non-adjacent"
+              << " swiftmodule at\n" << depPath << "\nin the trace.";
+          llvm::report_fatal_error(err.str());
+        }
+        buffer.clear();
+      }
+
+      // FIXME: Better error handling
+      StringRef realDepPath
+        = fs::real_path(depPath, buffer, /*expand_tile*/true)
+        ? StringRef(depPath) // Couldn't find the canonical path, assume
+                             // this is good enough.
+        : buffer.str();
+      buffer.clear();
+
+      traceInfo.push_back({
+        /*Name=*/
+        depMod->getName(),
+        /*Path=*/
+        realDepPath,
+        // TODO: There is an edge case which is not handled here.
+        // When we build a framework using -import-underlying-module, or an
+        // app/test using -import-objc-header, we should look at the direct
+        // imports of the bridging modules, and mark those as our direct
+        // imports.
+        /*IsImportedDirectly=*/
+        importedModules.find(depMod) != importedModules.end(),
+        /*SupportsLibraryEvolution=*/
+        depMod->isResilient()
+      });
+
+      continue;
+    }
+
+    // If the depTracker had an interface, that means that we must've
+    // built a swiftmodule from that interface, so we should have that
+    // filename available.
+    if (isSwiftinterface) {
+      err << "Unexpected path for swiftinterface file:\n" << depPath << "\n";
+      errorUnexpectedPath();
+    }
+
+    // Skip cached modules in the prebuilt cache. We will add the corresponding
+    // swiftinterface from the SDK directly, but this isn't checked. :-/
+    if (StringRef(depPath).contains(prebuiltCachePath))
+      continue;
+
+    // If we have a swiftmodule next to an interface, that interface path will
+    // be saved (not checked), so don't save the path to this swiftmodule.
+    SmallString<256> moduleAdjacentInterfacePath(depPath);
+    computeAdjacentInterfacePath(moduleAdjacentInterfacePath);
+    if (pathToModuleDecl.find(moduleAdjacentInterfacePath)
```

Comment:
You can get a slight simplification here with `count` instead of `find`.

Reply:
I think `find(..) != end()` is a fairly standard idiom, so I don't want to use `count != 0` here.

Reply:
`count` on sets is [just as standard](https://en.cppreference.com/w/cpp/container/set/count). If you don't want to use that, though, you could go with `llvm::is_contained`.

Reply:
> count on sets is just as standard.

What I meant by "standard idiom" is that `find() != end()` is a commonly understood idiom for `contains()`. Nothing complex about it. Replacing it with `pathToModuleDecl.count(..) != 0` feels more roundabout because we're not trying to _count_, we're trying to do a _find/lookup_.

Moreover, I have a map, not a set. This means that I cannot do `llvm::is_contained` because that needs a `DenseMapPair<StringRef, ModuleDecl *>`.